aws_lambda_powertools/utilities/data_masking/provider/kms/aws_encryption_sdk.py

```diff
@@ -0,0 +1,245 @@
+            max_bytes_encrypted=max_bytes_encrypted,
+            json_serializer=json_serializer,
+            json_deserializer=json_deserializer,
+        )
+
+    def encrypt(self, data: Any, provider_options: dict | None = None, **encryption_context: str) -> str:
+        return self._key_provider.encrypt(data=data, provider_options=provider_options, **encryption_context)
+
+    def decrypt(self, data: str, provider_options: dict | None = None, **encryption_context: str) -> Any:
+        return self._key_provider.decrypt(data=data, provider_options=provider_options, **encryption_context)
+
+
+class KMSKeyProvider:
+
+    """
+    The KMSKeyProvider is responsible for assembling an AWS Key Management Service (KMS)
+    client, a caching mechanism, and a keyring for secure key management and data encryption.
+    """
+
+    def __init__(
+        self,
+        keys: List[str],
+        json_serializer: Callable[..., str],
+        json_deserializer: Callable[[str], Any],
+        local_cache_capacity: int = CACHE_CAPACITY,
+        max_cache_age_seconds: float = MAX_CACHE_AGE_SECONDS,
+        max_messages_encrypted: int = MAX_MESSAGES_ENCRYPTED,
+        max_bytes_encrypted: int = MAX_BYTES_ENCRYPTED,
+    ):
+        session = botocore.session.Session()
+        register_feature_to_botocore_session(session, "data-masking")
+
+        self.json_serializer = json_serializer
+        self.json_deserializer = json_deserializer
+        self.client = EncryptionSDKClient()
+        self.keys = keys
+        self.cache = LocalCryptoMaterialsCache(local_cache_capacity)
+        self.key_provider = StrictAwsKmsMasterKeyProvider(key_ids=self.keys, botocore_session=session)
+        self.cache_cmm = CachingCryptoMaterialsManager(
+            master_key_provider=self.key_provider,
+            cache=self.cache,
+            max_age=max_cache_age_seconds,
+            max_messages_encrypted=max_messages_encrypted,
+            max_bytes_encrypted=max_bytes_encrypted,
+        )
+
+    def encrypt(self, data: Any, provider_options: dict | None = None, **encryption_context: str) -> str:
+        """
+        Encrypt data using the AWSEncryptionSDKProvider.
+
+        Parameters
+        -------
+            data : Union[bytes, str]
+                The data to be encrypted.
+            provider_options
+                Additional options for the aws_encryption_sdk.EncryptionSDKClient
+
+        Returns
+        -------
+            ciphertext : str
+                The encrypted data, as a base64-encoded string.
+        """
+        provider_options = provider_options or {}
+        self._validate_encryption_context(encryption_context)
+
+        data_encoded = self.json_serializer(data).encode("utf-8")
+
+        try:
+            ciphertext, _ = self.client.encrypt(
+                source=data_encoded,
+                materials_manager=self.cache_cmm,
+                encryption_context=encryption_context,
+                **provider_options,
```

Comment:
Is this supposed to be `provider_options` and `**encryption_context`? That is what I see in most places but in a few instances it is reversed.

Reply:
Because this is the actual API call, we pass the actual `dict` to the client. If we were to unpack `**encryption_context`, it would cause a SDK Client ParamError because it wouldn't have those key=value as accepted values. 

`**provider_options` on the other hand would be a 1:1 match with what's accepted by the client so it's safe to unpack - e.g., `provider_options={"algorithm": "..."}`